### ISSUE TITLE
Random hangs on "Startup script not finished yet. Waiting..." with GCP

### ISSUE THREAD

**Issue by zbjornson** [mitchellh/packer#3829]
- Packer Version: 0.10.2 (built from master, also occurring in 0.10.1)
- Host platform: Ubuntu 16.04
- Debug log output from `PACKER_LOG=1 packer build template.json` - does not appear to be useful, but I have if needed.

Somewhere around one in three times, packer gets stuck on 

```
==> googlecompute: Waiting for any running startup script to finish...
==> googlecompute: Startup script not finished yet. Waiting...
==> googlecompute: Startup script not finished yet. Waiting...
...
```

This always happens late in my build after several successful connections have already been made and most of the build is done. I don't know exactly when packer starts looking for this relative to running the build steps, but given that the file is called "step_wait_instance_startup.go" I would have assumed before the build. :confused:

I think this is because it's unreliable to rely on the serial port output for this. For all of the failed builds, the serial port output is missing the magic "Packer startup script done" text (which is what that step searches for). [Here's a gist with the serial output for one failed and one successful build](https://gist.github.com/zbjornson/d64a30aa9a39911e8a752327a1e68803). (Coincidentally some bytes are missing from the line where the metadata is set.) Another failed one had the magic text interrupted: 

```
Aug 24 18:44:28 ubuntu startupscript: Packer startup script starting.
Aug 24 18:44:29 ubuntu startupscript: Packer startup script Aug 24 18:44:29 ubuntu systemd[1]: Started Update UTMP about SysAug 24 18:44:30 ubuntu ntpd[2643]: Soliciting pool server 204.2.134.163
Aug 24 18:44:30 ubuntu ntpd[2643]: Soliciting pool server 173.255.196.241
```
(missing the "done.")

Could packer instead grep the startup script logs? That's /var/log/daemon.log (Debian, CentOS, RHEL images newer than v20160606) or /var/log/startupscript.log (Ubuntu; images older than v20160606) via ssh? ([docs](https://cloud.google.com/compute/docs/startupscript)) These are correct even when the serial output is not:

```
[host that is stuck]$ grep "startup script" /var/log/startupscript.log
Aug 24 18:44:28 ubuntu startupscript: Running startup script /var/run/google.startup.script
Aug 24 18:44:28 ubuntu startupscript: Packer startup script starting.
Aug 24 18:44:29 ubuntu startupscript: Packer startup script done.
Aug 24 18:44:29 ubuntu startupscript: Finished running startup script /var/run/google.startup.script
```

**Comment by rickard-von-essen** [comment]
What's your packer version?

**Comment by zbjornson** [comment]
0.10.2

**Comment by rickard-von-essen** [comment]
There is no such release. Latest is 0.10.1

**Comment by zbjornson** [comment]
Er, I suppose I built that from master about a week ago.

```
$ packer -v
0.10.2
```

This problem also happened in the prebuilt 0.10.1 though. Just haven't debugged it 'till now.

**Comment by rickard-von-essen** [comment]
If you do `make dev` the version gets suffixed with `.dev` if I remember correctly.

**Comment by rickard-von-essen** [comment]
Ok, also in 0.10.1

**Comment by rickard-von-essen** [comment]
Use `packer version` to get a detailed version specifier, very useful when building from source.

**Comment by zbjornson** [comment]
Thanks for the tip. `Packer v0.10.2.dev (0e334e67be81698a3630e9a01217ad0dc20609a0)`

I've also filed an issue with GCP because their serial output clearly has a problem, but the fix I suggested above might be faster to implement than their fix will come.

**Comment by zbjornson** [comment]
GCP has determined that the cause of the serial output issue is because the TTY buffer is being flushed during startup, potentially while the startup script is running. They're continuing to see *if* it can be fixed, but advised that looking at the startup logs over SSH (described in my OP) will be the most reliable option.

**Comment by rickard-von-essen** [comment]
I agree this is not robust. But I dislike to continue building in more logic into the builder that depends on the OS/distribution/version of the instance.  

You already mention a workaround that you can implement. Your last provisioner can be a small shell script that sleeps in a loop until it finds `Packer startup script done.` in the appropriate log file, or times out.

I don't eater understand why `StepWaitInstanceStartup` is after provisioning. I would expect it to be before `StepConnect` or at least before `StepProvision`.

@crunkleton maybe you can enlighten us about the ordering.

@zbjornson it would be interesting to experiment with just moving `StepWaitInstanceStartup` to before `StepConnect` and see if this is resolved. Also I think it's strange to run the _startup script_ in parallel with the _provisioners_.

**Comment by zbjornson** [comment]
> You already mention a workaround that you can implement. Your last provisioner can be a small shell script that sleeps in a loop until it finds Packer startup script done. in the appropriate log file, or times out.

I'm not sure how I could implement this as a work around without also modifying the packer source to stop it from looking for the startup script status itself?

I also doubt rearranging the steps (however logical) would fix this if the underlying data source is still corrupt. I am likewise interested to know why they seem to be out of order though.

Looking at the logs for Linux/BSD would be easy enough without adding logic (grep whichever file exists), but does packer do something similar for Windows?

**Comment by crunk1** [comment]
As for the ordering, I placed the startup script after the file provisioning because:

1. Placing `StepWaitInstanceStartup` after provisioning allows provisioning to run in parallel to the startup script running.
2. A user could want access to the files provisioned in their startup script.


On to the bug itself. Looking into it, it seems the getting the serial port output is unreliable. I'm sorry about that. You can assign this to me. Another solution would be to set a metadata variable on the instance upon completion.

**Comment by zbjornson** [comment]
Awesome, thanks. No need to apologize, I appreciate your contribs to an open-source project. The metadata variable idea is a smart one as far as avoiding OS-specific code.

**Comment by rickard-von-essen** [comment]
@crunk1 sounds great! Looking forward to a patch.

**Comment by jevonearth** [comment]
Sorry to bump a closed ticket.

I'm running into the same "Startup script not finished yet. Waiting..." issue while using packer, FreeBSD 11 and googlecompute. 

I have tried various metadata startup scripts, and startup_script_file settings, but I still get stuck with the `Waiting...` error. 

Could some one please clarify what users are expected to put on their packer config to get around this issue?

Thank you,

**Comment by rickard-von-essen** [comment]
@jevonearth can you open a new issue, I don't think this works correctly with FreeBSD and provide the requested info.

**Comment by crunk1** [comment]
Please CC me on the new issue.
